### PR TITLE
feat(ui): Add CompactSelect component 

### PR DESCRIPTION
**Pull request by vuluongj20**
`CompactSelect` is a wrapped `SelectControl` that provides more flexibility in terms of what the trigger button looks like. It follows a pattern for "popout" selects from [react-select's docs](https://react-select.com/advanced#experimental). 

This component is designed to replace `SelectControl` when there's limited space and a full-width select field wouldn't fit. In general, if a selector is outside of a form (which contains other form inputs), then we should use `CompactSelect`.

[Storybook Demo.](https://storybook-d5lzsrxr0.sentry.dev/?path=/story/components-compact-select--default)
[Sample implementation in `charts/optionSelector`.](https://github.com/getsentry/sentry/commit/afcc656b252ed5f1d35e46fca27d1af868896e13)

Screen Shot 2022-02-15 at 3 07 43 PM

### CODE REVIEW

File: static/app/components/forms/compactSelect.tsx [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+import space from 'sentry/styles/space';
+
+type TriggerProps = {
+  props: Omit<React.HTMLAttributes<Element>, 'children'> & {
+    onClick?: (e: MouseEvent) => void;
+  };
+  ref: React.RefObject<HTMLButtonElement>;
+};
+
+type Props = ControlProps &
```
Comment (JonasBa) [inline static/app/components/forms/compactSelect.tsx:30]: Same here, I think you actually want to use interface and extend the different props

```ts
interface Props extends ControlProps, Partial<OverlayProps>...

```

The main reason for that is that if you have an overlapping type, the union will merge the type definition and narrow down your type and may end up in never type.

Consider this overlap
```ts
type PropsA = {
    value: number | null
}

type PropsB = {
    value: string | boolean
}

type FinalProps = PropsA & PropsB

const props: FinalProps = {
    value: "test"
}
```

^^ will error out with `Type 'string' is not assignable to type 'never'.` whereas using an interface like

```ts
interface PropsA {
    value: number | null
}

interface PropsB {
    value: string | boolean
}

interface FinalProps extends PropsA, PropsB {}
```

Will properly error out with 
```
Interface 'FinalProps' cannot simultaneously extend types 'PropsA' and 'PropsB'.
 Named property 'value' of types 'PropsA' and 'PropsB' are not identical.(2320)
 ```

**Comment by vuluongj20** [comment]
@JonasBa hmm great points! made some changes in https://github.com/getsentry/sentry/pull/31850/commits/76553f62ac0b1f6dabd1d81a94a00f61e398ef2e. tell me if they look good!

File: static/app/components/forms/compactSelect.tsx [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+    const newValue = Array.isArray(option) ? option.map(opt => opt.value) : option?.value;
+    setInternalValue(newValue);
+    onChange?.(option);
+
+    if (closeOnSelect && !multiple) {
+      state.close();
+    }
+  }
+
+  function renderTrigger() {
```
Comment (JonasBa) [inline static/app/components/forms/compactSelect.tsx:199]: @vuluongj20 How do you feel about moving these to separate components or inlining them into the return call? This has a small benefit of making it easier to just read the render call and not have to jump back and forth between function return and renderTrigger call?

File: static/app/components/forms/compactSelect.tsx [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+        isOpen={state.isOpen}
+        {...triggerProps}
+        {...buttonProps}
+      >
+        {triggerLabel ?? label}
+      </DropdownButton>
+    );
+  }
+
+  function renderMenu() {
```
Comment (JonasBa) [inline static/app/components/forms/compactSelect.tsx:222]: Same inline comment here :)

Comment (vuluongj20) [reply static/app/components/forms/compactSelect.tsx:222]: 👆

File: static/app/components/forms/compactSelect.tsx (hunk 2 above)
Comment (JonasBa) [reply static/app/components/forms/compactSelect.tsx:199]: Yeah, I think using `!trigger ? <DropdownButton/> : null` is a super common react pattern (more common that having render* functions) so I would just go with that

**Comment by JonasBa** [comment]
@vuluongj20 I think there is a separate conversation to be had outside of the scope of this PR about coupling the presentation components with the logic of the dropdown. Imagine tomorrow that you need to create a new dropdown that looks close to this one, but with enough differences where it's required for you to make a new component. I assume that the way we would would do that is by copy pasting the styles of this menu and duplicating the presentational components and adding our custom logic - this means we now have two different menus to maintain from a design pov.

If we had separated presentational components, then we could just import our components with styles and add our custom business logic to them, leaving us with 1 menu to maintain from a user pov. This approach may seem verbose at first, but it allows you to refactor/change styles much faster in the future.

A quick example of something like this would be

```tsx
// CustomDropdown.tsx
function CustomDropDown(){
  <DropdownWrapper>
    <DropdownSearchInput {...customLogic} />
    <DropdownButton {...customButtonLogic}>1</DropDownButton>
    <DropdownButton {...customButtonLogic}>2</DropDownButton>
  </DropdownWrapper>
}

const DropdownWraper = styled("div")...
const DropdownSearchInput = styled("input")...
const DropdownButton = styled("button")...

// then in a new dropdown you would

// AnotherCustomDropdown.tsx
function AnotherCustomDropdown(){
  <DropdownWrapper>
    <DropdownSearchInput {...anotherCustomDropdownLogic} />
    <DropdownButton {...anotherCustomDropdownButtonLogic}>1</DropDownButton>
    <DropdownButton {...anotherCustomDropdownButtonLogic}>2</DropDownButton>
  </DropdownWrapper>
}

const DropdownWraper = styled("div")...
const DropdownSearchInput = styled("input")...
const DropdownButton = styled("button")...
```


But imagine if all we had to do was import the components like so

```tsx
// CustomDropdown.tsx
import {DropdownWrapper} from "sentry/components/DropdownWrapper"
import {DropdownSearchInput} from "sentry/components/DropdownSearchInput"
import {DropdownButton} from "sentry/components/DropdownButton"

function CustomDropDown(){
  <DropdownWrapper>
    <DropdownSearchInput {...customLogic} />
    <DropdownButton {...customButtonLogic}>1</DropDownButton>
    <DropdownButton {...customButtonLogic}>2</DropDownButton>
  </DropdownWrapper>
}

// AnotherCustomDropdown.tsx
import {DropdownWrapper} from "sentry/components/DropdownWrapper"
import {DropdownSearchInput} from "sentry/components/DropdownSearchInput"
import {DropdownButton} from "sentry/components/DropdownButton"

function AnotherCustomDropdown(){
  <DropdownWrapper>
    <DropdownSearchInput {...anotherCustomDropdownLogic} />
    <DropdownButton {...anotherCustomDropdownButtonLogic}>1</DropDownButton>
    <DropdownButton {...anotherCustomDropdownButtonLogic}>2</DropDownButton>
  </DropdownWrapper>
}
```

The main difference is that in the first example, the components are responsible for both the business logic and styling. This also leaves space for components to deviate far enough between styling that the users end up with an inconsistent UI experience. If tomorrow Sentry decides to change styling, we need to go find all occurrences of all custom dropdowns and fix them (which may not be trivial).

In the second example, our components are only responsible for the business logic and leave styling decisions to the UI library - if we want to update styles, there is a single place to do it. This approach is called atomic design and is it's a way to scale the codebase and keep it maintainable as it grows. I find building components like these a great exercise to writing unopinionated code that can easily be extended, but provides enough base to be functional. There is a great blog post from Brad Frost about atomic design on [this](https://bradfrost.com/blog/post/atomic-web-design/) topic.

Ideally, we should reach a point where engineers consuming the UI components never have to extend their styles :)

This is probably a topic of larger discussion though, but the natural followup to this would be writing hooks that encapsulate common behaviour and allow users to extend that business logic easily as well. I'm happy to jump on a call sometime if you want to bounce ideas about this or have any question :)

**Review by JonasBa — CHANGES_REQUESTED** [review]
@vuluongj20 sorry for being late to this, I suspect the acceptance test is a flake. I would  add some tests because there is a lot happening in the component body (mostly to test controlled/uncontrolled state)

**Comment by JonasBa** [comment]
@vuluongj20 is this still relevant, want me to give it another look?

**Comment by JonasBa** [comment]
Looks like @Jesse-Box included this component into our new designs. I'm ok if you dont want to add tests I guess, but I think you should include stories on how to use this components and all its different variations.

**Comment by vuluongj20** [comment]
@JonasBa Added a storybook story and also some tests for this. Tell me if they look good!

**Review by JonasBa — APPROVED** [review]
Sweet, thanks for adding the tests and stories 👏🏼

**Pull request merged**